persist/boltdb_test.go

```diff
@@ -149,14 +203,261 @@ func TestOpenDatabase(t *testing.T) {
+			t.Fatalf("closing a newly-emptied database for metadata %v, filename %v; error was %v", in.md, dbFilename, err)
 		}
 
 		// Clean up by deleting the testfile.
-		err = os.Remove(in.dbFilename)
+		err = os.Remove(dbFilepath)
 		if err != nil {
-			t.Fatalf("removing database file failing for input %v; error was %v", in, err)
+			t.Fatalf("removing database file failed for metadata %v, filename %v; error was %v", in.md, dbFilename, err)
 		}
 	}
+}
+
+// TestErrPermissionOpenDatabase tests calling OpenDatabase on a database file
+// with the wrong filemode (< 0600), which should result in an os.ErrPermission
+// error.
+func TestErrPermissionOpenDatabase(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
 
+	const (
+		dbHeader   = "Fake Header"
+		dbVersion  = "0.0.0"
+		dbFilename = "Fake Filename"
+	)
+
+	// Create a folder for the database file. If a folder by that name exists
+	// already, it will be replaced by an empty folder.
+	testDir := build.TempDir(persistDir, "TestErrPermissionOpenDatabase")
+	err := os.MkdirAll(testDir, 0700)
+	if err != nil {
+		t.Fatal(err)
+	}
+	dbFilepath := filepath.Join(testDir, dbFilename)
+	badFileModes := []os.FileMode{0000, 0001, 0002, 0003, 0004, 0005, 0010, 0040, 0060, 0105, 0110, 0126, 0130, 0143, 0150, 0166, 0170, 0200, 0313, 0470, 0504, 0560, 0566, 0577}
+
+	// Make sure OpenDatabase returns a permissions error for each of the modes
+	// in badFileModes.
+	for _, mode := range badFileModes {
+		// Create a file named dbFilename in directory testDir with the wrong
+		// permissions (mode < 0600).
+		_, err := os.OpenFile(dbFilepath, os.O_RDWR|os.O_CREATE, mode)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// OpenDatabase should return a permissions error because the database
+		// mode is less than 0600.
+		_, err = OpenDatabase(Metadata{dbHeader, dbVersion}, dbFilepath)
+		if !os.IsPermission(err) {
+			t.Errorf("OpenDatabase failed to return expected error when called on a database with the wrong permissions (%o instead of >= 0600);\n wanted:\topen %v: permission denied\n got:\t\t%v", mode, dbFilepath, err)
+		}
+
+		err = os.Remove(dbFilepath)
+		if err != nil {
+			t.Error(err)
+		}
+	}
+}
+
+// TestErrTxNotWritable checks that updateMetadata returns an error when called
+// from a read-only transaction.
+func TestErrTxNotWritable(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+	testDir := build.TempDir(persistDir, "TestErrTxNotWritable")
+	err := os.MkdirAll(testDir, 0700)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for i, in := range testInputs {
+		dbFilename := testFilenames[i%len(testFilenames)]
+		dbFilepath := filepath.Join(testDir, dbFilename)
+
+		db, err := bolt.Open(dbFilepath, 0600, &bolt.Options{Timeout: 3 * time.Second})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		boltDB := &BoltDatabase{
+			Metadata: in.md,
+			DB:       db,
+		}
+
+		err = db.View(func(tx *bolt.Tx) error {
```

Comment:
Generally better to favor fewer newlines in your code - more compact code means more fits on the screen, and if you are clumping ideas together well readability will improve.

I would generally pull out the newline here, and a couple of the others in this file.